config/common.php

```diff
@@ -22,12 +23,16 @@
     ],
     RuleHandlerResolverInterface::class => SimpleRuleHandlerContainer::class,
     'yii.validator.categorySource' => [
-        'definition' => static function (IdMessageReader $idMessageReader) use ($params): CategorySource {
-            return new CategorySource(
-                $params['yiisoft/validator']['translation.category'],
-                $idMessageReader,
-                extension_loaded('intl') ? new IntlMessageFormatter() : new SimpleMessageFormatter(),
-            );
+        'definition' => static function () use ($params): CategorySource {
+            $reader = class_exists(MessageSource::class)
+                ? new MessageSource(dirname(__DIR__) . '/messages')
+                : new IdMessageReader(); // @codeCoverageIgnore
```

Comment:
Maybe try to test it somehow? For example attempt to use an exception instead of `class_exists`.

Reply:
How use exception?

Reply:
Something like that.

In the config:

```php
try {
    new MessageSource();
} catch (\Error $exception) {
    $message = $exception->getMessage();
    if ($exception->getMessage() === 'Class "Yiisoft\Translator\Message\Php\MessageSource" not found') {
    // ...
    }
}
```

In the test (during building config):

```try
throw new \Error('Class "Yiisoft\Translator\Message\Php\MessageSource" not found');
```

Reply:
1) It's very difficult code for config. Current code very simple.
2) How in tests make `MessageSource` not available?

Reply:
I'll try to demonstrate in a separate PR.